Automattic/jetpack
# Add tsconfig presets for proper use of tsc and improve `jetpack generate`

## Proposed changes:

When setting `noEmit` to false for tsc, we need to make sure that `module` and `moduleResolution` are set to "nodenext" as well. Let's create some presets that do that, and set some other related configs.
    
Also let's clean up the base config a bit:
- We were setting `forceConsistentCasingInFileNames: true` everywhere, so let's go ahead and put that in the config.
- The values we had been setting for `lib` are already implied by setting `target` to "esnext" or "nodenext", so lets drop `lib`.
    
Then let's have `lint-project-structure.sh` complain if someone is trying to set these configs manually in tsconfig.json, since it's likely they won't get them all right.

And finally, let's update `jetpack generate` to do better for js-packages
    
- Ask whether it's JS or TS, and which build system (if any) is to be used.
- Use `src/index.jsx` instead of `src/index.js` by default.
- Set up webpack and/or ts deps if needed.
- Create `tsconfig.json` if TS is selected.
- Create `webpack.config.cjs` if Webpack is selected.
- Set up build scripts for the selected build system, if any.
- Don't bother asking about phpunit tests.
- Sort the resulting `scripts` and `devDependencies`.

### Other information:


## Jetpack product discussion


pdWQjU-FO-p2
Announcement post: pdWQjU-HO-p2

## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Is CI happy?
* Try reverting some of the tsconfig changes here and see if the updated `.github/files/lint-project-structure.sh` flags them.
* Try out `jetpack generate` for js-packages with various answers to the new question. Do the results look right?
* Check the built artifacts. Other than image-guide, which loses some Promise polyfills, nothing should be changing.
